File: benchmark/speedtest/cases/binary_trees/ref.rs
// The Rust counterpart of `main.fix`, on the same input, so the log can carry a
// reference the Fix line is read against. It checks the answer and prints nothing, as the
// Fix case does: a reference that computed something else would otherwise pass unnoticed.

// Allocation and reference-counting stress: build a full binary tree and count its nodes.
//
// Every node is heap-allocated, leaves included, so all three languages allocate 2^(n+1)-1
// times. A `Leaf` as its own enum variant would sit inside the parent's `Box` and allocate
// only for the internal nodes, which is half the work the other two do.
struct Tree {
    children: Option<(Box<Tree>, Box<Tree>)>,
}

fn make(d: i64) -> Box<Tree> {
    if d == 0 {
        Box::new(Tree { children: None })
    } else {
        Box::new(Tree {
            children: Some((make(d - 1), make(d - 1))),
        })
    }
}

fn check(t: &Tree) -> i64 {
    match &t.children {
        None => 1,
        Some((l, r)) => 1 + check(l) + check(r),
    }
}

fn main() {
    let n: i64 = 20;
    let t = make(n);
    assert_eq!(check(&t), 2_097_151);
}

File: benchmark/speedtest/cases/binary_trees/ref.c
// The C counterpart of `main.fix`, on the same input, so the log can carry a
// reference the Fix line is read against. It checks the answer and prints nothing, as the
// Fix case does: a reference that computed something else would otherwise pass unnoticed.

#include <stdio.h>
#include <stdlib.h>

typedef struct Tree { struct Tree *l, *r; } Tree;

static Tree *make(long d) {
    Tree *t = malloc(sizeof(Tree));
    if (d == 0) { t->l = NULL; t->r = NULL; }
    else { t->l = make(d - 1); t->r = make(d - 1); }
    return t;
}

static long check(Tree *t) {
    return t->l ? 1 + check(t->l) + check(t->r) : 1;
}

static void free_tree(Tree *t) {
    if (t->l) { free_tree(t->l); free_tree(t->r); }
    free(t);
}

int main(void) {
    long n = 20;
    Tree *t = make(n);
    long c = check(t);
    free_tree(t);
    if (c != 2097151) { fprintf(stderr, "binary_trees: %ld\n", c); return 1; }
    return 0;
}
